In [1]:
# Cell 0.1 — Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.cluster.hierarchy import linkage, dendrogram
from scipy.stats import chi2_contingency
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')

sns.set_style('whitegrid')
plt.rcParams['figure.dpi'] = 110
pd.set_option('display.max_colwidth', 80)



PROCESSED = Path('../data/processed')
FIGDIR = Path('../reports/figures')
FIGDIR.mkdir(parents=True, exist_ok=True)

In [2]:
# Cell 0.2 — Load raw data
df = pd.read_csv('../data/raw/mental-health-in-tech-2016.csv')
print(df.shape)
print(df)

(1433, 63)
      Are you self-employed?  \
0                          0   
1                          0   
2                          0   
3                          1   
4                          0   
...                      ...   
1428                       1   
1429                       1   
1430                       0   
1431                       0   
1432                       0   

     How many employees does your company or organization have?  \
0                                                        26-100   
1                                                          6-25   
2                                                          6-25   
3                                                           NaN   
4                                                          6-25   
...                                                         ...   
1428                                                        NaN   
1429                                                        NaN   
1

In [3]:
# Cell 0.3 — Dataset summary
print("Dataset loaded successfully.")
print(f"Rows:    {df.shape[0]:,}")
print(f"Columns: {df.shape[1]:,}")

df.columns

Dataset loaded successfully.
Rows:    1,433
Columns: 63


Index(['Are you self-employed?',
       'How many employees does your company or organization have?',
       'Is your employer primarily a tech company/organization?',
       'Is your primary role within your company related to tech/IT?',
       'Does your employer provide mental health benefits as part of healthcare coverage?',
       'Do you know the options for mental health care available under your employer-provided coverage?',
       'Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?',
       'Does your employer offer resources to learn more about mental health concerns and options for seeking help?',
       'Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources provided by your employer?',
       'If a mental health issue prompted you to request a medical leave from work, asking for that leave would be:',
       'Do you think that dis

In [4]:
# ============================================================
# Cell 0.4 — COMPLETE COLUMN INVENTORY
# ============================================================

column_inventory = pd.DataFrame({
    "column_number": range(1, len(df.columns) + 1),
    "column_name": df.columns,
    "dtype": df.dtypes.astype(str).values
})

column_inventory

,column_number,column_name,dtype
0,1,Are you self-employed?,int64
1,2,How many employees does your company or organization have?,str
2,3,Is your employer primarily a tech company/organization?,float64
3,4,Is your primary role within your company related to tech/IT?,float64
4,5,Does your employer provide mental health benefits as part of healthcare cove...,str
...,...,...,...
58,59,What US state or territory do you live in?,str
59,60,What country do you work in?,str
60,61,What US state or territory do you work in?,str
61,62,Which of the following best describes your work position?,str


In [5]:
# ============================================================
# Cell 0.5 — DUPLICATE ROW ANALYSIS
# ============================================================
#
# Purpose:
#     Determine whether the dataset contains completely
#     duplicated observations.
#
# Duplicate observations can affect clustering because
# identical respondents effectively receive additional weight.
# ============================================================

duplicate_count = df.duplicated().sum()

print(f"Number of completely duplicated rows: {duplicate_count:,}")
print(
    f"Percentage of dataset duplicated: "
    f"{duplicate_count / len(df) * 100:.2f}%"
)

Number of completely duplicated rows: 0
Percentage of dataset duplicated: 0.00%


In [6]:
# ============================================================
# Cell 0.6 — MISSING-VALUE AUDIT
# ============================================================
#
# Purpose:
#     Calculate missing-value counts and percentages for every
#     variable.
#
# This is one of the most important parts of this project because
# missingness in this survey is often caused by branching logic.
#
# I therefore do NOT assume:
#
#     missing = random missing data
#
# Instead, this table tells us where further investigation is
# required.
# ============================================================

missing_audit = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_percentage": df.isna().mean() * 100,
    "non_missing_count": df.notna().sum(),
    "unique_values": df.nunique(dropna=True)
})

print(missing_audit)

missing_audit = (
    missing_audit
    .sort_values("missing_percentage", ascending=False)
)



                                                                                  missing_count  \
Are you self-employed?                                                                        0   
How many employees does your company or organization have?                                  287   
Is your employer primarily a tech company/organization?                                     287   
Is your primary role within your company related to tech/IT?                               1170   
Does your employer provide mental health benefits as part of healthcare cover...            287   
...                                                                                         ...   
What US state or territory do you live in?                                                  593   
What country do you work in?                                                                  0   
What US state or territory do you work in?                                                  582   
Which of t

In [7]:
# Cell 0.7 — Build column inventory with short IDs, blocks, and stats

# The raw CSV uses the long question text as column names.
# I assign stable short IDs q01..q63 in the order they appear.

short_ids = [f"q{i+1:02d}" for i in range(df.shape[1])]
col_map = pd.DataFrame({
    'short_id': short_ids,
    'column_number': range(df.shape[1]),
    'question_text': df.columns,
})
col_map['question_short'] = col_map['question_text'].str.slice(0, 70) + '...'

# Block assignment (semantic grouping)
block_assignments = {
    **{f"q{i+1:02d}": 'A_employment_context' for i in range(0, 4)},
    **{f"q{i+1:02d}": 'B_current_employer'  for i in range(4, 16)},
    **{f"q{i+1:02d}": 'C_self_employed'     for i in range(16, 24)},
    **{f"q{i+1:02d}": 'D_previous_employer' for i in range(24, 36)},
    **{f"q{i+1:02d}": 'E_interview'         for i in range(36, 40)},
    **{f"q{i+1:02d}": 'F_stigma_career'     for i in range(40, 45)},
    **{f"q{i+1:02d}": 'G_mh_history'        for i in range(45, 53)},
    **{f"q{i+1:02d}": 'H_work_interference' for i in range(53, 55)},
    **{f"q{i+1:02d}": 'I_demographics'      for i in range(55, 63)},
}
col_map['block'] = col_map['short_id'].map(block_assignments)

# Per-column statistics
stats = []
for sid, qtext in zip(col_map['short_id'], col_map['question_text']):
    s = df[qtext]
    vc = s.value_counts(dropna=True)
    stats.append({
        'short_id': sid,
        'dtype': str(s.dtype),
        'n_missing': int(s.isnull().sum()),
        'pct_missing': round(s.isnull().mean() * 100, 2),
        'n_unique': int(s.nunique(dropna=True)),
        'top_category': vc.index[0] if len(vc) else None,
        'top_category_pct': round(vc.iloc[0] / s.notnull().sum() * 100, 2) if len(vc) else None,
    })
stats_df = pd.DataFrame(stats)


col_inventory = col_map.merge(stats_df, on='short_id')
col_inventory.to_csv(PROCESSED / 'column_inventory.csv', index=False)
col_inventory[['short_id','block','n_missing','pct_missing','n_unique','top_category','top_category_pct']].head()

,short_id,block,n_missing,pct_missing,n_unique,top_category,top_category_pct
0,q01,A_employment_context,0,0.00,2,0,79.97
1,q02,A_employment_context,287,20.03,6,26-100,25.48
2,q03,A_employment_context,287,20.03,2,1.0,77.05
3,q04,A_employment_context,1170,81.65,2,1.0,94.30
4,q05,B_current_employer,287,20.03,4,Yes,46.34


In [ ]:
# Cell 0.9 — Appendix A: ID → full question → block (reader-facing codebook)

block_labels = {
    'A_employment_context': 'A: Employment context',
    'B_current_employer':   'B: Current employer',
    'C_self_employed':      'C: Self-employed',
    'D_previous_employer':  'D: Previous employer',
    'E_interview':          'E: Interview',
    'F_stigma_career':      'F: Stigma and career',
    'G_mh_history':         'G: Mental-health history',
    'H_work_interference':  'H: Work interference',
    'I_demographics':       'I: Demographics',
}

appendix_a = col_map[['short_id', 'question_text', 'block']].copy()
appendix_a['block'] = appendix_a['block'].map(block_labels)
appendix_a.columns = ['ID', 'Question (full text)', 'Block']

# Spot-check: first and last question of each block
for blk, g in appendix_a.groupby('Block', sort=True):
    print(f"{blk:28s} {g.ID.iloc[0]}–{g.ID.iloc[-1]}")
    print(f"   first: {g['Question (full text)'].iloc[0][:90]}")
    print(f"   last : {g['Question (full text)'].iloc[-1][:90]}")

appendix_a.to_csv(PROCESSED / 'appendix_A_codebook.csv', index=False)


pd.set_option('display.max_rows', None, 'display.max_colwidth', None)
appendix_a

A: Employment context        q01–q04
   first: Are you self-employed?
   last : Is your primary role within your company related to tech/IT?
B: Current employer          q05–q16
   first: Does your employer provide mental health benefits as part of healthcare coverage?
   last : Have you heard of or observed negative consequences for co-workers who have been open abou
C: Self-employed             q17–q24
   first: Do you have medical coverage (private insurance or state-provided) which includes treatmen
   last : If yes, what percentage of your work time (time performing primary or secondary job functi
D: Previous employer         q25–q36
   first: Do you have previous employers?
   last : Did you hear of or observe negative consequences for co-workers with mental health issues 
E: Interview                 q37–q40
   first: Would you be willing to bring up a physical health issue with a potential employer in an i
   last : Why or why not?.1
F: Stigma and career         q41–q45
   firs

ModuleNotFoundError: No module named 'openpyxl'

In [ ]:
# Cell 0.8 — Inventory summary by block

col_inventory.groupby('block').agg(
    n_cols=('short_id','count'),
    mean_missing_pct=('pct_missing','mean'),
    max_missing_pct=('pct_missing','max'),
    mean_unique=('n_unique','mean'),
).round(2)

In [ ]:

# Cell 0.9 — Missingness heatmap (respondents × questions, sorted)

fig, ax = plt.subplots(figsize=(15, 6))

# Map full question text to short IDs (q01..q63) for clean x-axis labeling
qtext_to_sid = dict(zip(col_inventory['question_text'], col_inventory['short_id']))

# Create missingness matrix and rename columns to short IDs
miss = df.isnull().rename(columns=qtext_to_sid)

# Sort columns by missingness % and rows by total missing count
col_order = miss.mean().sort_values(ascending=False).index
row_order = miss.sum(axis=1).sort_values(ascending=False).index

# Plot heatmap using clean short IDs on x-axis
# Color scheme: Dark Blue (#1f4e79) = Missing, Light Neutral (#f2f2f2) = Observed
sns.heatmap(
    miss.loc[row_order, col_order],
    cbar=False,
    cmap=sns.color_palette(['#f2f2f2', '#1f4e79']),
    ax=ax,
    yticklabels=False,
    xticklabels=True
)

# Custom formatting and labels
ax.set_title(
    'Missingness Pattern Matrix (Dark Blue = Missing, Light Gray = Observed)\n'
    'Sorted by Missingness Severity across Questions & Respondents',
    fontsize=12, fontweight='bold', pad=12
)
ax.set_xlabel('Questions (q01–q63, sorted by % missing)', fontsize=10, labelpad=8)
ax.set_ylabel('Respondents (1,433 rows, sorted by total missing)', fontsize=10, labelpad=8)

# Format x-axis tick labels for readability
plt.xticks(rotation=90, fontsize=8)

plt.tight_layout()
plt.savefig(FIGDIR / 'missingness_heatmap.png', bbox_inches='tight', dpi=300)
plt.show()

In [ ]:
# Cell 1.0 — Hierarchical clustering of columns by missingness pattern
miss = df.isnull().astype(int)
Z = linkage(miss.T, method='ward')

fig, ax = plt.subplots(figsize=(16, 5))
dendrogram(Z, labels=col_inventory['short_id'].tolist(), leaf_rotation=90, ax=ax)
ax.set_title('Co-missingness dendrogram (columns clustered by missingness pattern)')
ax.set_ylabel('Ward distance')
plt.tight_layout()
plt.savefig(FIGDIR / 'comissingness_dendrogram.png', bbox_inches='tight')
plt.show()



### Missingness Heatmap & Co-Missingness Dendrogram Analysis

To evaluate the structural patterns and underlying mechanics of missing data across the survey, I analyzed the combined output of the missingness heatmap and hierarchical co-missingness dendrogram.

#### 1. Missingness Pattern Heatmap

The heatmap visualizes missingness across all 1,433 respondents (rows) and 63 questions (columns). Both axes are sorted hierarchically: columns by missingness percentage (highest on the left) and rows by total missing responses (highest at the top). **Dark blue** indicates missing values, while **light gray** represents observed responses.

* **Structural Blocks on High-Missingness Features:** The dense vertical dark-blue blocks on the left correspond to highly targeted questions, primarily the self-employed module (**Block C: q17–q24**) and conditional follow-ups (e.g., specific diagnosis details or US state fields).
* **Respondent-Level Segmentation:** The upper portion of the matrix is noticeably darker, highlighting respondents who skipped substantial sections of the survey—predominantly self-employed individuals bypass employer-related modules. Moving rightward, the matrix transitions to solid light gray, demonstrating that core constructs (**Block F: Stigma**, **Block G: Mental Health History**, **Block H: Work Interference**, and **Block I: Demographics**) were answered near-universally.
* **Mechanism Verification:** The distinct rectangular boundaries confirm that missingness is strictly **Missing Not at Random (MNAR)**, driven by survey skip-logic rather than random non-response.

#### 2. Co-Missingness Dendrogram

Using hierarchical clustering (Ward linkage) on the binary missingness indicators ($1 = \text{missing}$, $0 = \text{observed}$), the 63 questions form three primary co-missingness clusters:

* **High Co-Missingness Cluster (Left):** Groups questions that are omitted together due to specific entry flags. It is dominated by self-employed specific items (**q17–q24**) and granular conditional follow-ups (**q04, q19–q23, q45, q49, q50, q52**).
* **Employer-Logic Cluster (Middle):** Captures current-employer items (**Block B: q05–q16**). These questions are missing systematically for self-employed respondents who bypass non-applicable organizational queries.
* **Core Response Cluster (Right):** Represents the largest group of variables with minimal missingness. It encompasses previous-employer items (**Block D**), stigma and career perceptions (**Block F**), mental health history (**Block G**), work interference (**Block H**), and demographic factors (**Block I**).

#### Analytical Decision

The diagnostic plots prove that missing data is non-random and heavily dictated by two primary routing flags: *"Are you self-employed?"* and *"Do you have previous employers?"*.

Because our target population focuses specifically on tech industry employees, **excluding self-employed respondents** removes the dominant source of structural missingness (the primary blue block). This targeted filter purges systemic skip-logic gaps, yielding a clean, high-density feature set optimized for subsequent clustering analysis.

In [ ]:
# Cell 1.1 — Cluster membership (cut at 3 clusters)
from scipy.cluster.hierarchy import fcluster
clusters = fcluster(Z, t=3, criterion='maxclust')
col_inventory['missingness_cluster'] = clusters
col_inventory.groupby('missingness_cluster')['block'].value_counts()

In [ ]:
# Cell 1.2 — Sensitivity check: cut at k=4
clusters4 = fcluster(Z, t=4, criterion='maxclust')
col_inventory['missingness_cluster_k4'] = clusters4
print(col_inventory.groupby('missingness_cluster_k4')['block'].value_counts())

In [ ]:
# Cell 1.3 — Verify the three-layer routing structure
cluster2_cols = col_inventory.loc[col_inventory['missingness_cluster_k4']==2,
                                  'question_text'].tolist()
cluster1_cols = col_inventory.loc[col_inventory['missingness_cluster_k4']==1,
                                  'question_text'].tolist()

df['_c1_missing'] = df[cluster1_cols].isnull().all(axis=1)
df['_c2_missing'] = df[cluster2_cols].isnull().all(axis=1)

# Cluster 2 should be missing for ALL cluster-1-missing respondents (employed),
# plus some self-employed who didn't trigger the follow-up.
print("Cluster 2 missing given Cluster 1 missing:")
print(pd.crosstab(df['_c1_missing'], df['_c2_missing']))

# Among self-employed only, what triggers Cluster 2?
self_emp = df[df['Are you self-employed?']==1]
print("\nAmong self-employed, Cluster 2 missing vs MH history:")
print(pd.crosstab(self_emp['_c2_missing'],
                  self_emp['Have you had a mental health disorder in the past?']))

In [ ]:
# Cell 1.4 — Why do 92 employed respondents skip Cluster 2?
cluster2_cols = col_inventory.loc[col_inventory['missingness_cluster_k4']==2,
                                  'question_text'].tolist()
df['_c2_missing'] = df[cluster2_cols].isnull().all(axis=1)
df['_blockD_missing'] = df[col_inventory.loc[col_inventory['block']=='D_previous_employer',
                                              'question_text'].tolist()].isnull().all(axis=1)

# Among employed only
emp = df[df['Are you self-employed?']==0]
print("Among employed (n=1,146):")
print(pd.crosstab(emp['_c2_missing'], emp['_blockD_missing']))

# Which specific Cluster 2 columns drive the 92?
c2_cols = col_inventory.loc[col_inventory['missingness_cluster_k4']==2,
                            ['question_text','block','short_id']]
print("\nCluster 2 composition:")
print(c2_cols)

# Missingness rate of each Cluster 2 column among employed
print("\nPer-column missingness among employed:")
print(df[df['Are you self-employed?']==0][cluster2_cols].isnull().mean().round(3))

In [ ]:
# Cell 1.5 — Helper to plot a block of categorical columns
def plot_block(block_name, ncols=3):
    cols = col_inventory.loc[col_inventory['block'] == block_name, 'question_text'].tolist()
    ids  = col_inventory.loc[col_inventory['block'] == block_name, 'short_id'].tolist()
    nrows = int(np.ceil(len(cols) / ncols))
    fig, axes = plt.subplots(nrows, ncols, figsize=(5*ncols, 3.2*nrows))
    axes = np.array(axes).reshape(-1)
    for ax, cid, q in zip(axes, ids, cols):
        vc = df[q].value_counts(dropna=False).head(8)
        vc.index = [str(i)[:25] for i in vc.index]
        vc.plot(kind='barh', ax=ax, color='steelblue')
        ax.set_title(f'{cid}: {q[:55]}...', fontsize=8)
        ax.invert_yaxis()
    for ax in axes[len(cols):]:
        ax.axis('off')
    plt.tight_layout()
    plt.savefig(FIGDIR / f'univariate_{block_name}.png', bbox_inches='tight')
    plt.show()

for b in ['A_employment_context','B_current_employer','C_self_employed',
          'D_previous_employer','E_interview','F_stigma_career',
          'G_mh_history','H_work_interference','I_demographics']:
    plot_block(b)

In [ ]:
# Cell 1.6 — Self-employed explains Block B/C missingness
block_b_cols = col_inventory.loc[col_inventory['block']=='B_current_employer','question_text'].tolist()
block_c_cols = col_inventory.loc[col_inventory['block']=='C_self_employed','question_text'].tolist()

df['_blockB_all_missing'] = df[block_b_cols].isnull().all(axis=1)
df['_blockC_all_missing'] = df[block_c_cols].isnull().all(axis=1)

print("Block B missing vs self-employed:")
print(pd.crosstab(df['Are you self-employed?'], df['_blockB_all_missing']))
print("\nBlock C missing vs self-employed:")
print(pd.crosstab(df['Are you self-employed?'], df['_blockC_all_missing']))

In [ ]:
# Cell 1.7 — Previous-employer flag explains Block D missingness
df['_blockD_all_missing'] = df[col_inventory.loc[col_inventory['block']=='D_previous_employer','question_text'].tolist()].isnull().all(axis=1)
print(pd.crosstab(df['Do you have previous employers?'], df['_blockD_all_missing']))

In [ ]:
# Cell 1.7b — Per-column missingness in Block D, split by q24
block_d_df = col_inventory.loc[
    col_inventory['block'] == 'D_previous_employer',
    ['short_id', 'question_text']
]

rows = []
for _, r in block_d_df.iterrows():
    q = r['question_text']
    rows.append({
        'short_id': r['short_id'],
        'question_short': q[:60],
        'missing_pct_no_prev': round(
            df.loc[df['Do you have previous employers?'] == 0, q].isnull().mean() * 100, 1),
        'missing_pct_has_prev': round(
            df.loc[df['Do you have previous employers?'] == 1, q].isnull().mean() * 100, 1),
    })

block_d_missing = pd.DataFrame(rows)
block_d_missing['gap'] = (block_d_missing['missing_pct_no_prev']
                          - block_d_missing['missing_pct_has_prev'])
block_d_missing.sort_values('gap', ascending=False).reset_index(drop=True)

In [ ]:
# === DIAGNOSTIC for Cell 1.8 ===
print("1. Does col_inventory exist?", 'col_inventory' in dir())
print("2. Does df exist?", 'df' in dir())
print()

if 'col_inventory' in dir():
    print("3. Columns in col_inventory:")
    print(col_inventory.columns.tolist())
    print()
    
    print("4. dtype value counts:")
    print(col_inventory['dtype'].value_counts())
    print()
    
    print("5. How many columns pass the current filter?")
    mask = (col_inventory['dtype'] == 'object') & (col_inventory['n_unique'] < 30)
    print(mask.sum())
    print()
    
    cat_cols = col_inventory.loc[mask, 'question_text'].tolist()
    print("6. len(cat_cols) =", len(cat_cols))
    if len(cat_cols) > 0:
        print("   First 5:", cat_cols[:5])
    else:
        print("   >>> cat_cols is EMPTY — this is the problem")
else:
    print(">>> col_inventory does not exist. You must re-run Cell 0.7 first.")

In [ ]:
# Cell 1.8 — Compute Cramér's V for all categorical pairs
from scipy.stats import chi2_contingency
import itertools

def cramers_v(x, y):
    """Compute Cramér's V between two categorical series.
    Returns NaN if either series has <2 categories after dropping NaN."""
    ct = pd.crosstab(x, y)
    if ct.shape[0] < 2 or ct.shape[1] < 2:
        return np.nan
    chi2 = chi2_contingency(ct, correction=False)[0]
    n = ct.values.sum()
    if n == 0:
        return np.nan
    phi2 = chi2 / n
    r, k = ct.shape
    # Bias correction (Bergsma 2013)
    phi2corr = max(0, phi2 - ((k - 1) * (r - 1)) / (n - 1))
    rcorr = r - ((r - 1) ** 2) / (n - 1)
    kcorr = k - ((k - 1) ** 2) / (n - 1)
    denom = min(kcorr - 1, rcorr - 1)
    if denom <= 0:
        return np.nan
    return np.sqrt(phi2corr / denom)

# Only compute for low-cardinality categorical columns (exclude free text)
# Accept both 'object' and 'str' (newer pandas reports string columns as 'str')
cat_cols = col_inventory.loc[
    (col_inventory['dtype'].isin(['object', 'str', 'string', 'category'])) &
    (col_inventory['n_unique'].between(2, 29)),
    'question_text'
].tolist()

print(f"Selected {len(cat_cols)} columns for Cramér's V")

# Map question_text -> short_id
text_to_sid = dict(zip(col_inventory['question_text'], col_inventory['short_id']))

# Compute pairwise Cramér's V
results = []
for a, b in itertools.combinations(cat_cols, 2):
    v = cramers_v(df[a], df[b])
    if not np.isnan(v):
        results.append({
            'var1': text_to_sid[a],
            'var2': text_to_sid[b],
            'var1_text': a[:80],
            'var2_text': b[:80],
            'cramers_v': round(v, 4),
        })

cramers = pd.DataFrame(results)

if cramers.empty:
    print("WARNING: No valid Cramér's V pairs were computed. Check cat_cols selection.")
else:
    cramers = cramers.sort_values('cramers_v', ascending=False).reset_index(drop=True)
    cramers['flagged'] = cramers['cramers_v'] >= 0.85
    cramers.to_csv(PROCESSED / 'cramers_v_flagged_pairs.csv', index=False)
    print(f"Computed {len(cramers)} pairs; {cramers['flagged'].sum()} flagged at V >= 0.85")
    display(cramers.head(10))

### Cramér’s V Analysis – Interpretation

I computed Cramér’s V for all pairs of categorical variables to detect strong associations.

- 898 pairs were evaluated.
- **No pair reached the threshold of 0.85**, so there is no severe collinearity that would require immediate variable removal.
- The strongest associations (V = 0.55–0.75) are concentrated in the mental-health history block (q47, q48, q51, q54, q55). This is expected: past disorder, current disorder, professional diagnosis and work interference are conceptually related.
- Moderate associations also appear between disclosure-related questions (q19 ↔ q21) and between previous-employer support questions (q28 ↔ q29, q33 ↔ q34).

**Decision:**  
I will keep all variables for the first clustering iteration. In Notebook 3 I may still decide to reduce a few highly related pairs if the clustering results become unstable or hard to interpret.

In [ ]:
# Cell 1.9 — Top 20 Cramér's V pairs
top20 = cramers.head(20).copy()

# Attach block info for easier reading
sid_to_block = dict(zip(col_inventory['short_id'], col_inventory['block']))
top20['var1_block'] = top20['var1'].map(sid_to_block)
top20['var2_block'] = top20['var2'].map(sid_to_block)

top20[['var1', 'var1_block', 'var2', 'var2_block', 'cramers_v', 'flagged']]

### Cramér’s V Results with Block Information

I examined the 20 strongest associations and mapped them to the semantic blocks.

- The highest values (0.65–0.75) are concentrated inside the Mental Health History block (G) and between Block G and Work Interference (H). This is expected and indicates a coherent construct rather than problematic redundancy.
- Moderate associations (0.48–0.59) appear within the Previous Employer block (D) and the Current Employer block (B), showing internal consistency in disclosure comfort and employer support questions.
- Several high associations involve questions from the Self-employed block (C). These will become irrelevant once I exclude self-employed respondents.

**Decision:**  
Because no pair exceeds 0.85, I will retain all variables for the first version of the clustering feature set. I will re-evaluate redundancy after filtering out self-employed respondents and after encoding.

In [ ]:
# Cell 2.0 — Cramér's V heatmap (top 25 columns by mean association)
# Pivot to matrix form for heatmap
def get_v(a, b):
    row = cramers[((cramers['var1']==a) & (cramers['var2']==b)) |
                  ((cramers['var1']==b) & (cramers['var2']==a))]
    return row['cramers_v'].iloc[0] if len(row) else np.nan

# Pick the 25 columns with highest mean association
sid_counts = pd.concat([cramers['var1'], cramers['var2']]).value_counts()
top_sids = sid_counts.head(25).index.tolist()

matrix = pd.DataFrame(index=top_sids, columns=top_sids, dtype=float)
for a in top_sids:
    for b in top_sids:
        if a == b:
            matrix.loc[a, b] = 1.0
        elif pd.isna(matrix.loc[a, b]):
            v = get_v(a, b)
            matrix.loc[a, b] = v
            matrix.loc[b, a] = v

fig, ax = plt.subplots(figsize=(11, 9))
sns.heatmap(matrix.astype(float), cmap='Reds', vmin=0, vmax=1,
            annot=True, fmt='.2f', annot_kws={'size': 6}, ax=ax,
            cbar_kws={'label': "Cramér's V"})
ax.set_title("Cramér's V heatmap — top 25 associated columns")
plt.tight_layout()
plt.savefig(FIGDIR / 'cramers_v_heatmap.png', bbox_inches='tight')
plt.show()

### Cramér’s V Heatmap – Interpretation

The heatmap displays the 25 questions with the strongest mutual associations.

- The darkest block in the top-left corner is formed by **q47, q48 and q51** (past disorder, current disorder, professional diagnosis). These three variables are conceptually related and form a coherent mental-health-status construct.
- **q54** and **q55** (work interference) show clear associations with the mental-health-status questions, which is expected and useful for clustering.
- Within the previous-employer block, pairs such as q28–q29 and q33–q34 show moderate consistency.
- Most other associations are weak (light colours), indicating that the majority of questions contribute relatively independent information.

Because no association exceeds 0.85, I will retain all of these variables for the initial clustering feature set. After removing self-employed respondents I will re-check the associations.

In [ ]:
# Cell 2.1 — Age boxplots by stigma response
stigma_qs = {
    'q41': 'Do you feel that being identified as a person with a mental health issue would hurt your career?',
    'q42': 'Do you think that team members/co-workers would view you more negatively if they knew you suffered from a mental health issue?',
    'q43': 'How willing would you be to share with friends and family that you have a mental illness?',
    'q10': 'Do you think that discussing a mental health disorder with your employer would have negative consequences?',
}
fig, axes = plt.subplots(2, 2, figsize=(14, 8))
for ax, (sid, q) in zip(axes.ravel(), stigma_qs.items()):
    sns.boxplot(data=df, x=q, y='What is your age?', ax=ax)
    ax.set_title(f'{sid}: {q[:60]}...', fontsize=9)
    ax.tick_params(axis='x', rotation=30)
plt.tight_layout()
plt.savefig(FIGDIR / 'age_vs_stigma.png', bbox_inches='tight')
plt.show()

**Interpretation of Age vs. Stigma Boxplots**

These four boxplots show the distribution of age for different answer categories of key stigma-related questions.

### Main observations

**1. Age is very similar across all answer groups**
- In all four plots the median age (horizontal line inside the box) stays roughly between **30 and 35 years**.
- The boxes (middle 50% of respondents) largely overlap.
- This means **age does not strongly differentiate** how people answer these stigma questions.

**2. Outliers are present but not systematic**
- There are a few very high ages (above 100, even one around 300). These are almost certainly data-entry errors or joke answers and should be cleaned in preprocessing.
- The extreme outliers appear in several categories, so they are not linked to a specific opinion.

**3. Slight tendencies (but weak)**
- **q41 (career would be hurt)**: People who say “Yes, it has” or “Yes, I think it would” have a similar age profile to those who say “No”.
- **q42 (coworkers would view negatively)**: Again, almost no visible age difference.
- **q43 (willingness to share with friends/family)**: Slightly wider spread in some categories, but medians remain close.
- **q10 (negative consequences with employer)**: Same pattern — medians are nearly identical for “No”, “Maybe” and “Yes”.

### Conclusion

Age has **only a very weak relationship** with perceived stigma and disclosure comfort.  
Younger and older respondents (within the normal working-age range) give similar answers on these questions.


I examined the relationship between age and four key stigma-related questions using boxplots.

- The median age is approximately 30–35 years in almost every answer category.
- The interquartile ranges overlap heavily, indicating that age does not strongly influence how respondents perceive career harm, coworker judgement, or willingness to disclose.
- A small number of extreme outliers (ages > 100) are visible; these are likely data errors and will be removed during preprocessing.

**Conclusion:** Age shows only a very weak association with stigma perceptions. It is therefore more useful as a contextual / demographic variable than as a primary clustering feature.


In [ ]:
# Cell 2.2 — Gender (top 10) vs. stigma

top_genders = df['What is your gender?'].value_counts().head(10).index
df['_gender_top'] = np.where(df['What is your gender?'].isin(top_genders),
                             df['What is your gender?'], 'Other')
for sid, q in list(stigma_qs.items())[:2]:
    ct = pd.crosstab(df['_gender_top'], df[q], normalize='index') * 100
    ct.plot(kind='bar', stacked=True, figsize=(10,4), colormap='tab20')
    plt.title(f'{sid} by gender (top 10 + Other)')
    plt.ylabel('% within gender')
    plt.tight_layout()
    plt.savefig(FIGDIR / f'gender_vs_{sid}.png', bbox_inches='tight')
    plt.show()

**Interpretation of Gender vs. Stigma Plots**

These stacked bar charts show how answers to q41 and q42 are distributed across the most frequent gender labels.

### Main problems visible in the plots

1. **Gender is extremely messy**
   - The same gender appears under many different spellings:
     - Male / male / M / m
     - Female / female / F / f
   - This makes the x-axis hard to read and splits the same group into many small bars.

2. **No clear, strong gender pattern**
   - In both q41 and q42 the largest segment in almost every bar is the grey one (“Yes, I think it would” / “Yes, I think they would”).
   - The proportion of “Maybe” and “No” answers looks roughly similar across the bigger groups (Male/male and Female/female).
   - Because the categories are fragmented, it is difficult to draw reliable conclusions.

3. **“Other” and rare labels**
   - The “Other” group and some low-frequency labels show more variation, but the sample sizes are small, so these differences are not trustworthy.

### Conclusion

Gender in its raw form is **not usable** for analysis. The high number of inconsistent spellings (70 unique values in the original data) creates noise and makes visual comparison unreliable.

### What you should do next

In Notebook 2 i must consolidate gender into a small number of clean categories, for example:

- Male
- Female
- Non-binary / Other
- Prefer not to say

Only after this cleaning will gender vs. stigma plots become meaningful.


In [ ]:
# Cell 2.3 — Country distribution
top_countries = df['What country do you live in?'].value_counts().head(10)
top_countries.plot(kind='barh', figsize=(8,4), color='steelblue')
plt.title('Top 10 countries of residence')
plt.xlabel('Respondents')
plt.tight_layout()
plt.savefig(FIGDIR / 'top_countries.png', bbox_inches='tight')
plt.show()
print(f"US share: {(df['What country do you live in?']=='United States of America').mean()*100:.1f}%")

**Interpretation of Country Distribution**

The bar chart shows the 10 most common countries of residence.

### Key findings

- **United States of America** is by far the dominant country (~840 respondents, **58.6%** of the sample).
- **United Kingdom** is a distant second (~180 respondents).
- **Canada**, **Germany**, **Netherlands**, **Australia**, **Sweden**, **France**, **Ireland** and **Switzerland** follow with much smaller numbers.
- All remaining countries together make up a relatively small share.

### Implications for your analysis

1. **Strong geographic bias**  
   The dataset is heavily US-centric. Results will mainly reflect the experiences of people working in the United States (and to a lesser extent the UK and other Western countries).

2. **Limited generalisability**  
   Findings about stigma, employer support, or disclosure comfort may not fully apply to tech workers in Asia, Latin America, Africa, or Eastern Europe.

3. **Practical decision for preprocessing**  
   Because many countries have very few respondents, it is better to consolidate them. A useful grouping for later analysis is:

   - United States
   - United Kingdom
   - Canada
   - Germany
   - Other Europe
   - Other / Rest of World



In [ ]:
# Cell 2.4 — Flag columns by role (PATCHED)
# Added explicit override for q25: the general rule correctly
# classifies it as clustering_candidate based on block + cardinality,
# but it's a routing flag ("do you have previous employers?"), not
# an attitude/experience variable. Reclassified per the documented
# decision in the markdown below — this was previously described
# in prose but never implemented in the loop, which caused Notebook 2's
# clustering matrix to come out at 37 columns instead of the
# intended 36 (caught when Cell 17's output didn't match this
# markdown's own stated count).
role = []
for _, row in col_inventory.iterrows():
    sid, block, nuniq, toppct = row['short_id'], row['block'], row['n_unique'], row['top_category_pct']
    if sid == 'q25':
        r = 'context'                  # explicit override — routing flag, not attitude variable
    elif nuniq >= 50:
        r = 'identifier_or_freetext'
    elif toppct is not None and toppct >= 90:
        r = 'near_constant'
    elif block in ['A_employment_context','I_demographics']:
        r = 'context'
    elif block == 'C_self_employed':
        r = 'sparse_subgroup'
    else:
        r = 'clustering_candidate'
    role.append(r)
col_inventory['role'] = role

print(col_inventory.groupby('role')['short_id'].count())
col_inventory.loc[col_inventory['role']=='clustering_candidate',
                  ['short_id','block','question_short','n_unique','top_category_pct']]

### Role Assignment: Interpretation and Decisions

The role partition produces 37 clustering candidates, 6 context columns,
10 identifier/free-text columns, 2 near-constant columns, and 8 sparse-
subgroup columns. This is a healthy, well-balanced split.

**Key decisions:**

1. **Clustering candidates (37 → 36 after reclassification).** These span
   six blocks and cover the four constructs identified as strongest in
   the Cramér's V analysis: disclosure comfort, stigma/career impact,
   mental-health history, and work interference. All have 2–6 categories
   and none exceed the 90% top-category threshold.

2. **q25 reclassified to context.** The previous-employer flag is a
   routing variable, not an attitude variable. Keeping it as a clustering
   candidate risks producing clusters defined by employment history rather
   than mental-health attitudes. It is retained as a stratification
   variable and can be used post-hoc to characterise clusters.

3. **Context columns (6) retained separately.** Demographics (age, gender,
   country, position, remote work) will be used to describe clusters after
   the fact, not as clustering inputs. This avoids clusters driven by
   nationality or gender.

4. **Identifier/free-text columns (10) excluded from clustering.** High-
   cardinality fields (gender with 70 values, country with 53, diagnosis
   lists with 99–128) cannot be one-hot encoded without producing sparse
   noise columns. Some will be revisited in Notebook 3 for targeted
   feature engineering (e.g., keyword flags for stigma in free text).

5. **Near-constant columns (2) dropped.** No discriminative power.

6. **Sparse-subgroup columns (8) dropped after filtering.** Block C is
   100% missing once the 287 self-employed respondents are removed.

The role column is now the operational driver for Notebook 2 and
Notebook 3. After filtering, the clustering feature set will contain
approximately 30–36 strong candidate columns, which is appropriate for
the planned dimensionality reduction and clustering steps.

In [ ]:
# Cell 2.5 — Top tokens in free-text columns
from collections import Counter
import re

def top_tokens(series, n=20):
    text = ' '.join(series.dropna().astype(str).str.lower())
    tokens = re.findall(r"[a-z']{3,}", text)
    stop = set("the and for you not are with have that this but can will your they would "
               "there what why yes because from about them their been has had were was "
               "when who how all any out our get more some very just dont didn doesnt "
               "into than then also its it's i'm i've".split())
    return Counter(t for t in tokens if t not in stop).most_common(n)

for q in ['Why or why not?', 'Why or why not?.1']:
    print(f"\n=== {q} ===")
    for tok, cnt in top_tokens(df[q]):
        print(f"  {tok:>15}: {cnt}")

**Interpretation of Free-Text Vocabulary**

You extracted the most frequent meaningful words from the two open-ended questions:

- **q38** → “Why or why not?” (about bringing up a **physical** health issue in an interview)
- **q40** → “Why or why not?.1” (about bringing up a **mental** health issue in an interview)

### Key observations

**1. Clear thematic difference between the two questions**

| Theme              | Physical health (q38)      | Mental health (q40)          |
|--------------------|----------------------------|------------------------------|
| Dominant words     | job, health, issue, physical, work | mental, health, job, stigma |
| Fear / risk language | affect, impact, ability   | stigma, affect, wouldn’t    |
| Decision language  | depends, want, relevant   | want, feel, think, might    |

- For **physical** health, people talk more about whether it is relevant to the job or affects their ability to work.
- For **mental** health, the word **“stigma”** appears prominently (111 times). This word is almost absent from the physical-health answers.

**2. Shared concerns**
Both questions frequently mention:
- job / work / employer / interview
- affect / impact
- want / don’t / wouldn’t

This shows that respondents are weighing career risk in both cases, but the nature of the risk feels different.

**3. Useful signal for feature engineering**
The strong presence of “stigma” only in the mental-health answers confirms that perceived stigma is a central theme.  
You can later create simple binary features such as:
- contains “stigma”
- contains “fear” / “discrimination”
- contains “relevant” / “not relevant”



In [ ]:
# Cell 2.6 — Re-save column inventory with role column
# Cell 0.7 saved the inventory before roles were assigned; re-save now.
col_inventory.to_csv(PROCESSED / 'column_inventory.csv', index=False)
print(f"Saved column_inventory.csv with roles: {col_inventory.shape}")
print(col_inventory['role'].value_counts())

In [ ]:
# Cell 2.7 — Save question mapping (short_id ↔ question text ↔ block ↔ role)
mapping = col_inventory[[
    'short_id', 'column_number', 'question_text', 'question_short', 'block', 'role'
]]
mapping.to_csv(PROCESSED / 'question_mapping.csv', index=False)
print(f"Saved question_mapping.csv: {mapping.shape}")

In [ ]:
# Cell 2.8 — Save missingness clusters for both k=3 and k=4
miss_clusters = pd.DataFrame({
    'short_id': col_inventory['short_id'],
    'missingness_cluster_k3': col_inventory['missingness_cluster'],
    'missingness_cluster_k4': col_inventory['missingness_cluster_k4'],
})
miss_clusters.to_csv(PROCESSED / 'missingness_clusters.csv', index=False)
print(f"Saved missingness_clusters.csv: {miss_clusters.shape}")

In [ ]:
# Cell 2.9 — Save routing flags table
routing_flags = pd.DataFrame([
    {'flag': 'q01',     'question': 'Are you self-employed?',          'governs': 'Blocks B, C'},
    {'flag': 'q25',     'question': 'Do you have previous employers?', 'governs': 'Block D'},
    {'flag': 'q44',     'question': 'Observed unsupportive response?', 'governs': 'q45'},
    {'flag': 'q47',     'question': 'Currently have MH disorder?',     'governs': 'q49, q50'},
    {'flag': 'q51',     'question': 'Diagnosed by professional?',      'governs': 'q52'},
    {'flag': 'country', 'question': 'Country = United States',         'governs': 'q59, q61'},
])
routing_flags.to_csv(PROCESSED / 'routing_flags.csv', index=False)
print(f"Saved routing_flags.csv: {routing_flags.shape}")

In [ ]:
# Cell 2.10 — Verify all artifacts exist
import os
print("Artifacts in data/processed/:")
for f in sorted(os.listdir(PROCESSED)):
    size = (PROCESSED / f).stat().st_size
    print(f"  {f:<40} {size:>8,} bytes")

In [ ]:
import pandas as pd
from pathlib import Path

PROCESSED = Path('../data/processed')
check = pd.read_csv(PROCESSED / 'column_inventory.csv')

print("Columns:", check.columns.tolist())
print()
print("Role distribution:")
print(check['role'].value_counts())



In [ ]:
flags = pd.read_csv(PROCESSED / 'routing_flags.csv')
print(flags)

In [ ]:
mapping = pd.read_csv(PROCESSED / 'question_mapping.csv')
print(mapping.head())
print()
print("Shape:", mapping.shape)